File: Prime_numbers_factors_&_number_theory/Q1. Count Primes/Q1_CountPrimes.java
class Solution {
    public int countPrimes(int n) {
        // If n <= 2, there are no primes
        if (n <= 2) return 0;

        // Create a boolean array to track prime numbers
        boolean[] isPrime = new boolean[n];
        
        // Assume all numbers from 2 to n-1 are prime
        for (int i = 2; i < n; i++) {
            isPrime[i] = true;
        }

        // Use Sieve of Eratosthenes to find primes
        for (int i = 2; i * i < n; i++) {
            // If i is still marked as prime
            if (isPrime[i]) {
                // Mark all multiples of i as not prime
                for (int j = i * i; j < n; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        // Count how many numbers are still marked as prime
        int count = 0;
        for (int i = 2; i < n; i++) {
            if (isPrime[i]) count++;
        }

        // Return the total prime count
        return count;
    }
}

File: Prime_numbers_factors_&_number_theory/Q1. Count Primes/Q1_CountPrimes.py
class Solution(object):
    def countPrimes(self, n):
        """
        :type n: int
        :rtype: int
        """
        # If n is less than or equal to 2, there are no primes
        if n <= 2:
            return 0

        # Create a list marking all numbers as prime initially
        is_prime = [True] * n

        # 0 and 1 are not prime numbers
        is_prime[0] = is_prime[1] = False

        # Start checking from 2
        i = 2
        while i * i < n:
            # If i is still marked as prime
            if is_prime[i]:
                # Mark all multiples of i as not prime
                for j in range(i * i, n, i):
                    is_prime[j] = False
            # Move to next number
            i += 1

        # Count all numbers that are still prime
        return sum(is_prime)
